### ISSUE TITLE
Oxidized crashes after adding FTOS switch

### ISSUE THREAD

**Issue by sh0x** [ytti/oxidized#1898]
Hi, we have oxidized running fine for a few hundred devices running nx-os, brocade, bigip, cumulus, and arista.

I enabled FTOS in model_map and Oxidized starts fine, but when it collects the FTOS device it crashes.

I'm running Oxidized as root because I wasn't sure how to make it run as the oxidized user. All oxidized folders are owned by the oxidized user.

/var/log/messages
```Sep 10 17:33:38 us01-0-oxidized oxidized: Permission denied @ dir_s_mkdir - ~
Sep 10 17:33:38 us01-0-oxidized oxidized: /home/oxidized/.rbenv/versions/2.3.8/lib/ruby/2.3.0/fileutils.rb:254:in `mkdir': Permission denied @ dir_s_mkdir - ~ (Errno::EACCES)
Sep 10 17:33:38 us01-0-oxidized oxidized: from /home/oxidized/.rbenv/versions/2.3.8/lib/ruby/2.3.0/fileutils.rb:254:in `fu_mkdir'
Sep 10 17:33:38 us01-0-oxidized oxidized: from /home/oxidized/.rbenv/versions/2.3.8/lib/ruby/2.3.0/fileutils.rb:228:in `block (2 levels) in mkdir_p'
Sep 10 17:33:38 us01-0-oxidized oxidized: from /home/oxidized/.rbenv/versions/2.3.8/lib/ruby/2.3.0/fileutils.rb:226:in `reverse_each'
Sep 10 17:33:38 us01-0-oxidized oxidized: from /home/oxidized/.rbenv/versions/2.3.8/lib/ruby/2.3.0/fileutils.rb:226:in `block in mkdir_p'
Sep 10 17:33:38 us01-0-oxidized oxidized: from /home/oxidized/.rbenv/versions/2.3.8/lib/ruby/2.3.0/fileutils.rb:211:in `each'
Sep 10 17:33:38 us01-0-oxidized oxidized: from /home/oxidized/.rbenv/versions/2.3.8/lib/ruby/2.3.0/fileutils.rb:211:in `mkdir_p'
Sep 10 17:33:38 us01-0-oxidized oxidized: from /home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/node.rb:80:in `rescue in run_input'
Sep 10 17:33:38 us01-0-oxidized oxidized: from /home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/node.rb:66:in `run_input'
Sep 10 17:33:38 us01-0-oxidized oxidized: from /home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/node.rb:44:in `block in run'
Sep 10 17:33:38 us01-0-oxidized oxidized: from /home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/node.rb:38:in `each'
Sep 10 17:33:38 us01-0-oxidized oxidized: from /home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/node.rb:38:in `run'
Sep 10 17:33:38 us01-0-oxidized oxidized: from /home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/job.rb:9:in `block in initialize'
```

/home/oxidized/.config/oxidized/crash
```[oxidized@us01-0-oxidized oxidized]$ cat crash
--------------------------------------------------
2019-09-10 17:33:38 UTC
Permission denied @ dir_s_mkdir - ~ [Errno::EACCES]
--------------------------------------------------
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/2.3.0/fileutils.rb:254:in `mkdir'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/2.3.0/fileutils.rb:254:in `fu_mkdir'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/2.3.0/fileutils.rb:228:in `block (2 levels) in mkdir_p'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/2.3.0/fileutils.rb:226:in `reverse_each'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/2.3.0/fileutils.rb:226:in `block in mkdir_p'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/2.3.0/fileutils.rb:211:in `each'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/2.3.0/fileutils.rb:211:in `mkdir_p'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/node.rb:80:in `rescue in run_input'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/node.rb:66:in `run_input'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/node.rb:44:in `block in run'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/node.rb:38:in `each'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/node.rb:38:in `run'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/job.rb:9:in `block in initialize'
```

/home/oxidized/.config/oxidized/logs/log
```D, [2019-09-10T18:06:23.332854 #26561] DEBUG -- : lib/oxidized/worker.rb: Added blr-00/blr-00-r01msw01 to the job queue
D, [2019-09-10T18:06:23.332894 #26561] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2019-09-10T18:06:23.332962 #26561] DEBUG -- : lib/oxidized/job.rb: Starting fetching process for blr-00-r01msw01 at 2019-09-10 18:06:23 UTC
D, [2019-09-10T18:06:23.333090 #26561] DEBUG -- : lib/oxidized/input/ssh.rb: Connecting to blr-00-r01msw01
D, [2019-09-10T18:06:23.333168 #26561] DEBUG -- : AUTH METHODS::["none", "publickey", "password", "keyboard-interactive"]
D, [2019-09-10T18:06:24.333102 #26561] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2019-09-10T18:06:25.333602 #26561] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2019-09-10T18:06:26.333925 #26561] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2019-09-10T18:06:27.334226 #26561] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2019-09-10T18:06:27.966030 #26561] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^([\w.@-]+[#>]\s?)$/] at blr-00-r01msw01
D, [2019-09-10T18:06:28.334624 #26561] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2019-09-10T18:06:29.334956 #26561] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2019-09-10T18:06:29.709253 #26561] DEBUG -- : lib/oxidized/input/cli.rb: Running post_login commands at blr-00-r01msw01
D, [2019-09-10T18:06:29.709383 #26561] DEBUG -- : lib/oxidized/input/cli.rb: Running post_login command: "terminal length 0", block: nil at blr-00-r01msw01
D, [2019-09-10T18:06:29.709421 #26561] DEBUG -- : lib/oxidized/input/ssh.rb terminal length 0 @ blr-00-r01msw01 with expect: /^([\w.@-]+[#>]\s?)$/
D, [2019-09-10T18:06:29.709592 #26561] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^([\w.@-]+[#>]\s?)$/] at blr-00-r01msw01
D, [2019-09-10T18:06:30.335263 #26561] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2019-09-10T18:06:30.358599 #26561] DEBUG -- : lib/oxidized/input/cli.rb: Running post_login command: "terminal width 0", block: nil at blr-00-r01msw01
D, [2019-09-10T18:06:30.358722 #26561] DEBUG -- : lib/oxidized/input/ssh.rb terminal width 0 @ blr-00-r01msw01 with expect: /^([\w.@-]+[#>]\s?)$/
D, [2019-09-10T18:06:30.358905 #26561] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^([\w.@-]+[#>]\s?)$/] at blr-00-r01msw01
D, [2019-09-10T18:06:31.050664 #26561] DEBUG -- : lib/oxidized/model/model.rb Collecting commands' outputs
D, [2019-09-10T18:06:31.050829 #26561] DEBUG -- : lib/oxidized/model/model.rb Executing show inventory
D, [2019-09-10T18:06:31.050872 #26561] DEBUG -- : lib/oxidized/input/ssh.rb show inventory @ blr-00-r01msw01 with expect: /^([\w.@-]+[#>]\s?)$/
D, [2019-09-10T18:06:31.051051 #26561] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^([\w.@-]+[#>]\s?)$/] at blr-00-r01msw01
D, [2019-09-10T18:06:31.335555 #26561] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2019-09-10T18:06:31.729042 #26561] DEBUG -- : lib/oxidized/model/model.rb Executing show inventory media
D, [2019-09-10T18:06:31.729159 #26561] DEBUG -- : lib/oxidized/input/ssh.rb show inventory media @ blr-00-r01msw01 with expect: /^([\w.@-]+[#>]\s?)$/
D, [2019-09-10T18:06:31.729345 #26561] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^([\w.@-]+[#>]\s?)$/] at blr-00-r01msw01
D, [2019-09-10T18:06:32.335841 #26561] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2019-09-10T18:06:32.377393 #26561] DEBUG -- : lib/oxidized/model/model.rb Executing show running-config
D, [2019-09-10T18:06:32.377509 #26561] DEBUG -- : lib/oxidized/input/ssh.rb show running-config @ blr-00-r01msw01 with expect: /^([\w.@-]+[#>]\s?)$/
D, [2019-09-10T18:06:32.377682 #26561] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^([\w.@-]+[#>]\s?)$/] at blr-00-r01msw01
F, [2019-09-10T18:06:33.025556 #26561] FATAL -- : Oxidized crashed, crashfile written in /home/oxidized/.config/oxidized/crash

**Comment by litewhatever** [comment]
@sh0x can you test if this issue is a duplicate of https://github.com/ytti/oxidized/issues/1854

**Comment by sh0x** [comment]
Hi, I tried the patch in #1854 but my oxidized still crashes at the same place. I also tried removing line 41 from ftos.rb directly.

**Comment by litewhatever** [comment]
Interesting, whats the version you are running? Can you perhaps try to replicate those commands manually which is ran by ftos.rb to see where it crashes exactly?

**Comment by seros1521** [comment]
@sh0x 
Faced the same problem.

1. An exception occurs during data collection (in my case, an unprintable character appeared in snmp-location)
2. oxidized tries to write stacktrace to a file in the directory specified by the `crash.directory` parameter.
3. I used the wildcard character "~" in this parameter, but oxidized does not open it, but tries to use it as an absolute path, as expected, does not find it and exits.

So check your configuration for the use of wildcards. Parameters are processed differently, in some special characters are expanded, in some they are not. In the oxidized source code, the `crash.directory` parameter is configured as
https://github.com/ytti/oxidized/blob/17cdbc8c302cd6041d3373bf4d66f98538337304/lib/oxidized/config.rb#L38
And the value of `Oxidized :: Config :: Root`:
https://github.com/ytti/oxidized/blob/17cdbc8c302cd6041d3373bf4d66f98538337304/lib/oxidized/config.rb#L6
Check the value of these environment variables `OXIDIZED_HOME` and` HOME` (if you set them yourself, if not, they must have the correct values).

After that, an error during data collection should be written to the `crash.directory` directory, and oxidized will continue to work. And you can separately investigate and correct this error.

**Comment by sh0x** [comment]
Right on, thanks @seros1521 ! I had the crash directory defined in the config file with '~'. After changing it to the full path, and oxidized doesn't crash now, and writes output to the crashes directory.

Unfortunately I can't find the issue in the config causing the crash. Our switch config is fairly generic. I removed all interface descriptions and configurations with strings but it still errors.

@litewhatever below is the output from log and crashes directory in case you have any hints, getting NoMethodError.

`D, [2020-04-07T16:50:10.365229 #15830] DEBUG -- : lib/oxidized/model/model.rb Executing show running-config
D, [2020-04-07T16:50:10.365367 #15830] DEBUG -- : lib/oxidized/input/ssh.rb show running-config @ blr-00-r01msw01 with expect: /^([\w.@-]+[#>]\s?)$/
D, [2020-04-07T16:50:10.365586 #15830] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^([\w.@-]+[#>]\s?)$/] at blr-00-r01msw01
E, [2020-04-07T16:50:10.982831 #15830] ERROR -- : 10.112.121.3 raised NoMethodError with msg "undefined method `join' for nil:NilClass", 10.112.121.3 saved`

`[oxidized@us01-0-oxidized oxidized]$ cat crashes/10.112.121.3
2020-04-07 16:42:44 UTC
undefined method `join' for nil:NilClass [NoMethodError]
--------------------------------------------------
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/model/ftos.rb:30:in `block in <class:FTOS>'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/model/model.rb:114:in `instance_exec'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/model/model.rb:114:in `cmd'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/model/model.rb:152:in `block in get'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/model/model.rb:151:in `each'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/model/model.rb:151:in `get'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/input/cli.rb:14:in `get'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/node.rb:67:in `run_input'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/node.rb:44:in `block in run'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/node.rb:38:in `each'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/node.rb:38:in `run'
/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/job.rb:9:in `block in initialize'`

**Comment by seros1521** [comment]
Error in line:
https://github.com/ytti/oxidized/blob/17cdbc8c302cd6041d3373bf4d66f98538337304/lib/oxidized/model/ftos.rb#L30
`to_a [3 ..- 1]` can return `nil` if the input string with the response from the device is too short. This can happen if the device does not understand the command, or another error occurs when the device processes the command and the response is shorter than expected. Check this by enabling debugging:
https://github.com/ytti/oxidized/blob/master/docs/Configuration.md#debugging

**Comment by seros1521** [comment]
You can also add debug output by line 30 to `/home/oxidized/.rbenv/versions/2.3.8/lib/ruby/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/model/ftos.rb`:
```
Oxidized.logger.error 'cfg1: %s' % [cfg]
Oxidized.logger.error 'cfg2: %s' % [cfg.each_line.to_a]
Oxidized.logger.error 'cfg3: %s' % [cfg.each_line.to_a[3..-1]]
```
and see the application log.

**Comment by sh0x** [comment]
Thanks! I enabled debugging as you suggested and was able to capture the command output from the switch.

`blr-00-r01msw01>show running-config
                     ^
% Error: Invalid input at "^" marker.`

So our new FTOS switch wasn't configured to drop directly into privileged exec mode like all our other gear. It works fine when I add enable secret to oxidized or configure FTOS to put the user directly into privileged exec mode.

Thanks for the help @seros1521  ! I appreciate the detailed information to troubleshoot oxidized and for identifying the reason oxidized was crashing due to crashlog path in config file.
